File: lib/data/datasources/remote_media_datasource.dart
import 'package:dio/dio.dart';

import '../../core/constants/api_constants.dart';
import '../../core/network/dio_client.dart';
import '../models/media_item_model.dart';

/// Lấy danh sách media (image + video) từ API standee.
class RemoteMediaDataSource {
  RemoteMediaDataSource({Dio? dio}) : _dio = dio ?? DioClient.I.client;

  final Dio _dio;

  /// GET /api/standee/version → { "version": 7 }
  Future<int> fetchVersion() async {
    try {
      final res = await _dio.get(ApiConstants.versionPath);
      final data = res.data;
      if (data is Map && data['version'] != null) {
        return (data['version'] as num).toInt();
      }
      return 0;
    } catch (_) {
      return 0;
    }
  }

  Future<List<MediaItem>> fetchMediaList() async {
    final res = await _dio.get(ApiConstants.playlistPath);
    final data = res.data;
    final list = _extractPosterList(data);
    return list
        .whereType<Map<String, dynamic>>()
        .map((e) => _mapPosterToMedia(e))
        .whereType<MediaItem>()
        .toList();
  }

  List<dynamic> _extractPosterList(dynamic value, [int depth = 0]) {
    if (value == null) return const [];
    if (value is List) return value;
    if (value is! Map) return const [];
    if (depth > 4) return const [];

    for (final key in const ['result', 'data', 'items', 'posters', 'activePosters']) {
      final v = value[key];
      if (v is List) return v;
      if (v is Map) {
        final nested = _extractPosterList(v, depth + 1);
        if (nested.isNotEmpty) return nested;
      }
    }

    for (final v in value.values) {
      if (v is List) return v;
      if (v is Map) {
        final nested = _extractPosterList(v, depth + 1);
        if (nested.isNotEmpty) return nested;
      }
    }
    return const [];
  }

  MediaItem? _mapPosterToMedia(Map<String, dynamic> json) {
    String? url = json['posterURL'] as String? ??
        json['url'] as String? ??
        json['mediaUrl'] as String? ??
        json['imageUrl'] as String? ??
        json['src'] as String?;

    final media = json['media'];
    if (url == null && media is Map) {
      url = media['url'] as String? ??
          media['mediaUrl'] as String? ??
          media['src'] as String?;
    }
    if (url == null || url.isEmpty) return null;

    if (!url.startsWith('http://') && !url.startsWith('https://') && !url.startsWith('asset:')) {
      final base = ApiConstants.baseUrl.replaceAll(RegExp(r'/$'), '');
      final path = url.startsWith('/') ? url : '/$url';
      url = '$base$path';
    }

    String typeStr = (json['type'] ?? '').toString().toLowerCase();
    if (typeStr.isEmpty) {
      if (url.endsWith('.mp4') || url.endsWith('.webm')) {
        typeStr = 'video';
      } else {
        typeStr = 'image';
      }
    }

    final type = typeStr == 'video' ? MediaType.video : MediaType.image;
    final duration =
        (json['duration'] ?? json['displayDuration'] ?? json['durationSeconds']) as int? ?? 10;
    final id = (json['id'] ?? json['posterId'] ?? '').toString();

    return MediaItem(
      id: id,
      url: url,
      type: type,
      durationSeconds: duration > 0 ? duration : 10,
    );
  }
}

File: src/data/datasources/remoteMediaDataSource.ts
import { getHttpClient } from '../../core/network/httpClient';
import { apiConstants } from '../../core/constants/apiConstants';
import { MediaItem, mediaItemFromJson } from '../models/MediaItem';

function extractPosterList(value: unknown, depth = 0): unknown[] {
  if (value == null) return [];
  if (Array.isArray(value)) return value;
  if (typeof value !== 'object') return [];
  if (depth > 4) return [];

  const obj = value as Record<string, unknown>;
  for (const key of ['result', 'data', 'items', 'posters', 'activePosters']) {
    const v = obj[key];
    if (Array.isArray(v)) return v;
    if (v && typeof v === 'object') {
      const nested = extractPosterList(v, depth + 1);
      if (nested.length > 0) return nested;
    }
  }
  for (const v of Object.values(obj)) {
    if (Array.isArray(v)) return v;
    if (v && typeof v === 'object') {
      const nested = extractPosterList(v, depth + 1);
      if (nested.length > 0) return nested;
    }
  }
  return [];
}

function mapPosterToMedia(json: Record<string, unknown>): MediaItem | null {
  let url: string | undefined =
    (json.posterURL as string) ??
    (json.url as string) ??
    (json.mediaUrl as string) ??
    (json.imageUrl as string) ??
    (json.src as string);

  const media = json.media as Record<string, unknown> | undefined;
  if (!url && media && typeof media === 'object') {
    url = (media.url as string) ?? (media.mediaUrl as string) ?? (media.src as string);
  }
  if (!url || url.length === 0) return null;

  const base = apiConstants.baseUrl.replace(/\/$/, '');
  if (!url.startsWith('http://') && !url.startsWith('https://') && !url.startsWith('asset:')) {
    url = url.startsWith('/') ? `${base}${url}` : `${base}/${url}`;
  }

  let typeStr = String(json.type ?? '').toLowerCase();
  if (!typeStr) {
    typeStr = url.endsWith('.mp4') || url.endsWith('.webm') ? 'video' : 'image';
  }

  const duration =
    (json.duration as number) ??
    (json.displayDuration as number) ??
    (json.durationSeconds as number) ??
    10;
  const id = String(json.id ?? json.posterId ?? '');

  return mediaItemFromJson({
    id,
    url,
    type: typeStr === 'video' ? 'video' : 'image',
    durationSeconds: duration > 0 ? duration : 10,
  });
}

export async function fetchVersion(): Promise<number> {
  try {
    const res = await getHttpClient().get<{ version?: number }>(apiConstants.versionPath);
    const data = res.data;
    if (data?.version != null) return Number(data.version);
    return 0;
  } catch {
    return 0;
  }
}

export async function fetchMediaList(): Promise<MediaItem[]> {
  return fetchPosters(apiConstants.standeeId);
}

/** Lấy danh sách poster active cho standee: GET /api/poster/{standeeId}/withActivePosters */
export async function fetchPosters(standeeId: string): Promise<MediaItem[]> {
  const path = apiConstants.getPosterPath(standeeId);
  const res = await getHttpClient().get(path);
  const list = extractPosterList(res.data);
  return list
    .filter((e): e is Record<string, unknown> => e != null && typeof e === 'object')
    .map(mapPosterToMedia)
    .filter((m): m is MediaItem => m != null);
}
